File: configuration/src/main/java/com/synopsys/integration/configuration/source/SpringConfigurationPropertySource.java
package com.synopsys.integration.configuration.source;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.function.BiConsumer;

import org.jetbrains.annotations.NotNull;
import org.springframework.boot.context.properties.source.ConfigurationProperty;
import org.springframework.boot.context.properties.source.ConfigurationPropertyName;
import org.springframework.boot.context.properties.source.ConfigurationPropertySource;
import org.springframework.boot.context.properties.source.ConfigurationPropertySources;
import org.springframework.boot.context.properties.source.InvalidConfigurationPropertyNameException;
import org.springframework.boot.context.properties.source.IterableConfigurationPropertySource;
import org.springframework.boot.env.RandomValuePropertySource;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.core.env.ConfigurablePropertyResolver;

import com.synopsys.integration.common.util.Bds;

public class SpringConfigurationPropertySource implements PropertySource {
    private final String name;
    private final IterableConfigurationPropertySource propertySource;
    private final ConfigurablePropertyResolver configurablePropertyResolver;

    public SpringConfigurationPropertySource(String name, IterableConfigurationPropertySource propertySource, ConfigurablePropertyResolver configurablePropertyResolver) {
        this.name = name;
        this.propertySource = propertySource;
        this.configurablePropertyResolver = configurablePropertyResolver;
    }

    public static List<SpringConfigurationPropertySource> fromConfigurableEnvironmentSafely(
        ConfigurableEnvironment configurableEnvironment,
        BiConsumer<String, Exception> unknownConsumer
    ) {
        try {
            return new ArrayList<>(fromConfigurableEnvironment(configurableEnvironment, false));
        } catch (RuntimeException e) {
            unknownConsumer.accept("An unknown property source was found, will ignore unknown property source and proceed.", e);
            return new ArrayList<>(fromConfigurableEnvironment(configurableEnvironment, true));
        }
    }

    // TODO - this should return an Optional
    public static List<SpringConfigurationPropertySource> fromConfigurableEnvironment(ConfigurableEnvironment configurableEnvironment, boolean ignoreUnknown) {
        List<ConfigurationPropertySource> sources = Bds.listOf(ConfigurationPropertySources.get(configurableEnvironment));
        return Bds.of(sources).map(it -> {
            if (IterableConfigurationPropertySource.class.isAssignableFrom(it.getClass())) {
                return getPropertySource(configurableEnvironment, ignoreUnknown, it);
            } else if (RandomValuePropertySource.class.isAssignableFrom(it.getUnderlyingSource().getClass())) {
                //We know an underlying random source can't be iterated but we don't care. It can't give a list of known keys.
                return null;
            } else {
                if (ignoreUnknown) {
                    return null;
                } else {
                    throw new RuntimeException(
                        new UnknownSpringConfigurationException(
                            "Unknown spring configuration type. We may be unable to find property information from it correctly. Likely a new configuration property source should be tested against. "
                        ));
                }
            }
        }).filterNotNull().toList();

    }

    // TODO - this should return an Optional
    private static SpringConfigurationPropertySource getPropertySource(
        ConfigurableEnvironment configurableEnvironment,
        boolean ignoreUnknown,
        ConfigurationPropertySource configurationPropertySource
    ) {
        Object underlying = configurationPropertySource.getUnderlyingSource();
        if (org.springframework.core.env.PropertySource.class.isAssignableFrom(underlying.getClass())) {
            org.springframework.core.env.PropertySource springSource = (org.springframework.core.env.PropertySource) underlying;
            return new SpringConfigurationPropertySource(springSource.getName(), (IterableConfigurationPropertySource) configurationPropertySource, configurableEnvironment);
        } else {
            if (ignoreUnknown) {
                return null;
            } else {
                throw new RuntimeException(
                    new UnknownSpringConfigurationException(
                        "Unknown underlying spring configuration source. We may be unable to determine where a property originated. Likely a new property source type should be tested against."));
            }
        }
    }

    @NotNull
    private Optional<ConfigurationPropertyName> toConfigurationName(String key) {
        try {
            return Optional.of(ConfigurationPropertyName.of(key));
        } catch (InvalidConfigurationPropertyNameException e) {
            return Optional.empty();
        }
    }

    private Optional<ConfigurationProperty> toConfigurationProperty(String key) {
        return toConfigurationName(key).map(name -> propertySource.getConfigurationProperty(name));
    }

    @Override
    public Boolean hasKey(String key) {
        Optional<ConfigurationPropertyName> configurationPropertyName = toConfigurationName(key);
        return configurationPropertyName.filter(propertyName -> propertySource.getConfigurationProperty(propertyName) != null).isPresent();
    }

    @Override
    public Set<String> getKeys() {
        return Bds.of(propertySource).map(ConfigurationPropertyName::toString).toSet();
    }

    @Override
    //Spring resolves configuration properties using a configurable property resolver (resolves --prop=${VAR} replaces VAR from an environment variable of the same name.)
    //The value provided from the configurable property itself or the property source is not resolved, so we have to get the final value from a resolver.
    //Theoretically, this resolver should be bound to the property source, but this current method at least gets the final resolved value. - jp
    public String getValue(String key) {
        return toConfigurationProperty(key)
            .map(ConfigurationProperty::getName)
            .map(ConfigurationPropertyName::toString)
            .map(configurablePropertyResolver::getProperty)
            .map(Object::toString)
            .orElse(null);
    }

    @Override
    public String getOrigin(String key) {
        return toConfigurationProperty(key).map(ConfigurationProperty::getOrigin).map(Object::toString).orElse(null);
    }

    @Override
    public String getName() {
        return name;
    }
}

File: configuration/src/main/java/com/synopsys/integration/configuration/source/SpringConfigurationPropertySource.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.source

import org.springframework.boot.context.properties.source.ConfigurationPropertyName
import org.springframework.boot.context.properties.source.ConfigurationPropertySources
import org.springframework.boot.context.properties.source.InvalidConfigurationPropertyNameException
import org.springframework.boot.context.properties.source.IterableConfigurationPropertySource
import org.springframework.boot.env.RandomValuePropertySource
import org.springframework.core.env.ConfigurableEnvironment

class UnknownSpringConfiguration(msg: String) : Exception(msg)

class SpringConfigurationPropertySource(private val name: String, private val propertySource: IterableConfigurationPropertySource) : PropertySource {
    companion object {
        fun fromConfigurableEnvironmentSafely(configurableEnvironment: ConfigurableEnvironment): List<SpringConfigurationPropertySource> {
            return fromConfigurableEnvironment(configurableEnvironment, true);
        }

        @Throws(UnknownSpringConfiguration::class)
        fun fromConfigurableEnvironment(configurableEnvironment: ConfigurableEnvironment, ignoreUnknown: Boolean = true): List<SpringConfigurationPropertySource> {
            val sources = ConfigurationPropertySources.get(configurableEnvironment).toList()
            return sources.map {
                if (it is IterableConfigurationPropertySource) {
                    val underlying = it.underlyingSource
                    if (underlying is org.springframework.core.env.PropertySource<*>) {
                        SpringConfigurationPropertySource(underlying.name, it)
                    } else {
                        if (ignoreUnknown) null else throw UnknownSpringConfiguration("Unknown underlying spring configuration source. We may be unable to determine where a property originated. Likely a new property source type should be tested against.")
                    }
                } else if (it.underlyingSource is RandomValuePropertySource) {
                    //We know an underlying random source can't be iterated but we don't care. It can't give a list of known keys.
                    null
                } else {
                    if (ignoreUnknown) null else throw UnknownSpringConfiguration("Unknown spring configuration type. We may be unable to find property information from it correctly. Likely a new configuration property source should be tested against. ")
                }
            }.filterNotNull()
        }
    }

    private fun toConfigurationName(key: String): ConfigurationPropertyName? {
        return try {
            ConfigurationPropertyName.of(key)
        } catch (e: InvalidConfigurationPropertyNameException) {
            null
        }
    }

    override fun hasKey(key: String): Boolean {
        return toConfigurationName(key)?.let { propertySource.getConfigurationProperty(it) } != null
    }

    override fun getValue(key: String): String? {
        return toConfigurationName(key)?.let { propertySource.getConfigurationProperty(it)?.value?.toString() }
    }

    override fun getName(): String = name

    override fun getOrigin(key: String): String? {
        return toConfigurationName(key)?.let { propertySource.getConfigurationProperty(it)?.origin?.toString() }
    }

    override fun getKeys(): Set<String> {
        return propertySource.map { it.toString() }.toSet()
    }
}